### DISCUSSION TITLE
extract Keyword break my grammar

### DISCUSSION THREAD

**Discussion by Tudyx** [tree-sitter/tree-sitter#2711] [Q&A] [answered]
While experimenting for a general tree-sitter grammar for log files, I've come to the point where I want to extract nodes whether there are within parentheses or not.
For instance, for the given log, I want to extract all the apples:

```
apple (apple)
```
A simple grammar that does exactly what I want is this one:
```javascript
 rules: {
    log_file: $ => repeat(
      choice(
        $.apple,
        $.apple_parenthesis,
      )
    ),
    apple: $ => "apple",
    apple_parenthesis: $ => seq('(', $.apple, ')'),
 }
```

That's great but if add some word with "apple" in them, they also get selected. For instance, the "bad_apple" in this log get also selected:

```
apple (apple) bad_apple
```

If I understand well, the way to emulate "word_bound"/ `\b` using tree sitter is to use the keyword extraction feature. So I modify the grammar to take this in account

```javascript
word: $ => $.word,

rules: {
    log_file: $ => repeat(
      choice(
        $.apple,
        $.apple_parenthesis,
        $.word,
      )
    ),
    apple: $ => "apple",
    apple_parenthesis: $ => seq('(', $.apple, `)`),
    word: $ => /\S+/,
 }
```

With this `bad_apple` doesn't get selected anymore but so do `(apple)` :cry: .  `(apple)`  get parsed as a word
```
(log_file [0, 0] - [1, 0]
  (apple [0, 0] - [0, 5])
  (word [0, 6] - [0, 13])
  (word [0, 14] - [0, 23]))
```

As `word` is the last rules of my grammar, I expect it to have the lowest priority. I've tried to play with `pre()` or `conflicts` but without success.

Obviously, something like
```javascript
choice("apple", "(apple)"),
```
will select what I want but my goal is to access the inner apple for syntax highlighting .

Please could you help me to select only the good apples?

**Comment by WillLillis** [comment, marked as answer]
Hi! I spent some a little time working on this and I think I figured it out :). I think the issue was your use of `\S` for `word`, as that seems to grab the parentheses in `(apple)`. I used a regex for `word` below to match any non-whitespace character besides '(' and ')', but you could also potentially use something like ` word: $ => /[a-zA-Z0-9_]+/`.

```js
module.exports = grammar({
    name: 'apple',

    word: $ => $.word,

    rules: {
        log_file: $ => repeat(
            choice(
                $.apple,
                $.apple_parens,
                $.word
            )
        ),

        apple: $ => "apple",
        apple_parens: $ => seq('(', $.apple, `)`),
        word: $ => /[^\s()]+/
    }
});
```

![image](https://github.com/tree-sitter/tree-sitter/assets/87077023/46ee1640-d894-4807-967c-0ff93f3b889e)

**Comment by Tudyx** [comment]
Thanks for your help! I've also ended up with the same solution on my side, see [this grammar](https://github.com/Tudyx/tree-sitter-log/blob/62cfe307e942af3417171243b599cc7deac5eab9/grammar.js#L143)

**Reply by WillLillis** [reply]
Nice!
